### PR TITLE
Add OAuth2 support & Use provided host for certificate verification

### PR DESCRIPTION
**Pull request by skiz**
Add OAuth2 support
Use provided host for certificate verification
Remove redundant ANONYMOUS mechanism support

### CODE REVIEW

File: xmpp.go [hunk 1]
```diff
@@ -172,15 +185,12 @@ func (o Options) NewClient() (*Client, error) {
 		if o.TLSConfig != nil {
 			tlsconn = tls.Client(c, o.TLSConfig)
 		} else {
-			DefaultConfig.ServerName = strings.Split(o.User, "@")[1]
+			DefaultConfig.ServerName = host
```
Comment (skiz) [inline xmpp.go]: This was the culprit causing TLS verification errors.  We should be using the provided host name and not parsing it from the username, as they can (and do) differ from the host.

File: xmpp.go [hunk 2]
```diff
@@ -315,19 +327,16 @@ func (c *Client) init(o *Options) error {
 
 		mechanism := ""
 		for _, m := range f.Mechanisms.Mechanism {
-			if m == "ANONYMOUS" {
```
Comment (skiz) [inline xmpp.go]: ANONYMOUS is already handled above when there is no user and password set.  This should be the correct and distinct way to connect.

**Comment by skiz** [comment]
go fmt cleans up additional files.

**Comment by mattn** [comment]
Thank you

**Pull request merged**